### ISSUE TITLE
Brainstorming projects to include

### ISSUE THREAD

**Issue by TomasVotruba** [TomasVotruba/awesome-php-ast#1]
* https://github.com/staabm/phpstan-dba
* https://github.com/phpmyadmin/sql-parser
* https://github.com/nikic/PHP-Parser
* https://github.com/matthiasnoback/phpstan-twig-analysis
* https://github.com/TomasVotruba/barista
* https://github.com/TomasVotruba/html-parser
* https://github.com/revealphp/reveal
* https://github.com/canvural/phpstan-blade-rule

**Comment by matthiasnoback** [comment]
Books: 

- https://leanpub.com/rector-the-power-of-automated-refactoring (Rector & AST-based manipulation)
- https://leanpub.com/recipes-for-decoupling (PHPStan & AST-based analysis)

:)

**Comment by arnaud-lb** [comment]
* https://github.com/phpstan/phpstan-doctrine uses the AST of Doctrine DQL queries to infer the type of their results
* https://github.com/arnaud-lb/php-sema uses an AST-based CFG for flow analysis (a wip project)
* https://github.com/ircmaxell/php-cfg creates an IR/TAC-based CFG from an AST

**Comment by TomasVotruba** [comment]
@ajthinking Thanks for the tips! Could you link the Doctrine DQL AST exactly? I could not find it. I don't want to copy every extension of every tool, as it would make the whole list cluttered and hard to read, but the DQL parser could "tool" on its own under "database" section.

**Comment by TomasVotruba** [comment]
@ajthinking As for https://github.com/ircmaxell/php-cfg, I read the README and its stop right at the moment of explaining what it does. Could you extend the README first with e.g. example of output for given input?
The [composer stats](https://packagist.org/packages/ircmaxell/php-cfg/stats) suggest it is rather niche project, so I'm not sure it would be useful here.

**Comment by TomasVotruba** [comment]
Thank you all for the inputs. I've put them to README markdown formated and kicked of :) 
https://github.com/TomasVotruba/awesome-php-ast/commit/e5d7a3442849609d7f44b3511b2845a10daf17db

For next ones, please send a pull-request to propperty category, in an "awesome list" spirit :+1:

**Comment by arnaud-lb** [comment]
@TomasVotruba I'm assuming that you wanted to mention me instead of @ajthinking :D

> Thanks for the tips! Could you link the Doctrine DQL AST exactly? I could not find it. I don't want to copy every extension of every tool, as it would make the whole list cluttered and hard to read, but the DQL parser could "tool" on its own under "database" section.

The Doctrine DQL AST is [here](https://github.com/doctrine/orm/tree/e8472c8/lib/Doctrine/ORM/Query/AST) (it represents [this grammar](https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/dql-doctrine-query-language.html#ebnf)), and in phpstan-doctrine it is analyzed mainly by [this class](https://github.com/phpstan/phpstan-doctrine/blob/5556305/src/Type/Doctrine/Query/QueryResultTypeWalker.php).

In terms of functionality, this is similar to phpstan-dba except that phpstan-doctrine works on Doctrine DQL while phpstan-dba works on SQL.

Basically, phpstan-doctrine allows PHPStan to know the precise type of $results here:

```
$results = $entityManager->createQuery('SELECT f.foo, f.bar FROM Foo f')->getResult()
```

> As for https://github.com/ircmaxell/php-cfg, I read the README and its stop right at the moment of explaining what it does. Could you extend the README first with e.g. example of output for given input?
> The [composer stats](https://packagist.org/packages/ircmaxell/php-cfg/stats) suggest it is rather niche project, so I'm not sure it would be useful here.

This is possibly too low level to appear here, indeed.

**Comment by TomasVotruba** [comment]
Yes, autocomplete was faster than me :D

Thanks for explaining 👍 

So the Doctrine PHPStan package is designed also to be used outside the extension? Do you have any docs/post about it? Direct PHP code without docs might be quite hard to understand.
